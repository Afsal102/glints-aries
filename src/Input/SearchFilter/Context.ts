import { createContext, useContext } from 'react';

interface SearchFilterContext {
  closeSearchFilterMenu: () => void;
}

export const SearchFilterContext = createContext<
  SearchFilterContext | undefined
>(undefined);

export function useSearchFilterContext() {
  const context = useContext(SearchFilterContext);
  if (context === undefined) {
    throw new Error('useSearchFilterContext must be used within SearchFilter');
  }
  return context;
}

export const MainHeaderHeightContext = createContext(0);
