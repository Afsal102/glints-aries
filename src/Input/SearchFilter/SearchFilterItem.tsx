import * as React from 'react';
import { SearchFilterItemWrapper } from './SearchFilterStyle';
import { useSearchFilterContext } from './Context';

const SearchFilterItem: React.FunctionComponent<Props> = props => {
  const { children, className, onClick, ...defaultProps } = props;
  const { closeSearchFilterMenu } = useSearchFilterContext();

  const handleClick = (e: React.MouseEvent<HTMLLIElement, MouseEvent>) => {
    if (onClick) {
      onClick(e);
    }
    if (closeSearchFilterMenu) {
      closeSearchFilterMenu();
    }
  };
  return (
    <SearchFilterItemWrapper
      className={className}
      onClick={handleClick}
      role="listitem"
      {...defaultProps}
    >
      {children}
    </SearchFilterItemWrapper>
  );
};

export interface Props
  extends React.ComponentPropsWithoutRef<typeof SearchFilterItemWrapper> {
  children: React.ReactNode;
  onClick?(event: React.MouseEvent<HTMLLIElement, MouseEvent>): void;
}

export default SearchFilterItem;
