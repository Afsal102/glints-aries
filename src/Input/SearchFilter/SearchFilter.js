/* @flow */

import React, { Component, Fragment } from 'react';

import { SearchFilterContainer, SearchFilterBar, SearchFilterBody } from '../../Style/Input/SearchFilterStyle';

class SearchFilter extends Component <Props, State> {
  state = {
    isOpen: false,
  };

  handleOpen = () => {
    const { isOpen } = this.state;
    this.setState({ isOpen: !isOpen });
  };

  render() {
    const {
      children,
      className,
      label,
      content,
      value,
      ...defaultProps
    } = this.props;

    const { isOpen } = this.state;

    return (
      <SearchFilterContainer
        id="aries-searchfilter"
        className={className}
        role="search"
        aria-expanded={isOpen}
        aria-label={label}
      >
        <SearchFilterBar id="searchfilter-inputwrapper">
          <input
            type="text"
            placeholder={label}
            onFocus={this.handleOpen}
            onBlur={this.handleOpen}
            value={value}
            {...defaultProps}
          />
          {content}
        </SearchFilterBar>
        <SearchFilterBody
          id="searchfilter-content"
          role="menuitem"
          aria-hidden={!isOpen && true}
          open={isOpen}
        >
          { children }
        </SearchFilterBody>
      </SearchFilterContainer>
    );
  }
}

type Props = {
  children: React$Node,
  content: React$Node,
  className: string,
  label: string,
  value: string,
};

type State = {
  isOpen: boolean,
};

export default SearchFilter;
