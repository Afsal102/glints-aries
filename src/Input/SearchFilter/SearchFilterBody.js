/* @flow */

import React from 'react';
import { SearchFilterResultContainer } from '../../Style/Input/SearchFilterStyle';

const SearchFilterBody = (props: Props) => {
  const {
    children,
    className,
    ...defaultProps
  } = props;

  return (
    <SearchFilterResultContainer
      className={className}
      role="menuitem"
      aria-hidden={false}
      {...defaultProps}
    >
      {children}
    </SearchFilterResultContainer>
  );
};

type Props = {
  children: React$Node,
  className: string,
};

export default SearchFilterBody;
