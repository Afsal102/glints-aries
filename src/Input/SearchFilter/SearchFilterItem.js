/* @flow */

import React from 'react';
import { SearchFilterItemWrapper } from '../../Style/Input/SearchFilterStyle';

const SearchFilterItem = (props: Props) => {
  const {
    children,
    className,
    onClick,
    ...defaultProps
  } = props;

  return (
    <SearchFilterItemWrapper
      className={className}
      onMouseDown={onClick}
      {...defaultProps}
    >
      {children}
    </SearchFilterItemWrapper>
  );
};

type Props = {
  children: React$Node,
  onClick: Function,
  className: string,
};

export default SearchFilterItem;
