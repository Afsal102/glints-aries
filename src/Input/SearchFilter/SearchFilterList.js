/* @flow */

import React from 'react';
import { SearchFilterListWrapper } from '../../Style/Input/SearchFilterStyle';

const SearchFilterList = (props: Props) => {
  const {
    title,
    children,
    className,
    ...defaultProps
  } = props;

  return (
    <SearchFilterListWrapper className={className} {...defaultProps}>
      <label>
        {title}
      </label>
      <ul>
        {children}
      </ul>
    </SearchFilterListWrapper>
  );
};

type Props = {
  children: React$Node,
  className: string,
  title: string,
};

export default SearchFilterList;
