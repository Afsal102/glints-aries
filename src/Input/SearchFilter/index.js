// @flow

import SearchFilter from './SearchFilter';
import SearchFilterBody from './SearchFilterBody';
import SearchFilterList from './SearchFilterList';
import SearchFilterItem from './SearchFilterItem';

SearchFilter.Body = SearchFilterBody;
SearchFilter.List = SearchFilterList;
SearchFilter.Item = SearchFilterItem;

export default SearchFilter;
