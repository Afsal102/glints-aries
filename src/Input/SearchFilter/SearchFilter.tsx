import * as React from 'react';

import { SearchFilterContainer, SearchFilterBar, SearchFilterBodyWrapper } from '../../Style/Input/SearchFilterStyle';
import SearchFilterBody from './SearchFilterBody';
import SearchFilterList from './SearchFilterList';
import SearchFilterItem from './SearchFilterItem';

class SearchFilter extends React.Component<Props, State> {
  static Body = SearchFilterBody;
  static List = SearchFilterList;
  static Item = SearchFilterItem;

  state = {
    isOpen: false,
  };

  handleOpen = () => {
    const { isOpen } = this.state;
    this.setState({ isOpen: !isOpen });
  };

  render() {
    const {
      children,
      className,
      label,
      content,
      value,
      ...defaultProps
    } = this.props;

    const { isOpen } = this.state;

    return (
      <SearchFilterContainer
        id="aries-searchfilter"
        className={className}
        role="search"
        aria-expanded={isOpen}
        aria-label={label}
      >
        <SearchFilterBar id="searchfilter-inputwrapper">
          <input
            type="text"
            placeholder={label}
            onFocus={this.handleOpen}
            onBlur={this.handleOpen}
            value={value}
            {...defaultProps}
          />
          {content}
        </SearchFilterBar>
        <SearchFilterBodyWrapper
          id="searchfilter-content"
          role="menuitem"
          aria-hidden={!isOpen && true}
          open={isOpen}
        >
          { children }
        </SearchFilterBodyWrapper>
      </SearchFilterContainer>
    );
  }
}

interface Props extends Omit<React.HTMLProps<HTMLInputElement>, 'content'> {
  children: React.ReactNode;
  content: React.ReactNode;
  label: string;
};

interface State {
  isOpen: boolean;
}

export default SearchFilter;
