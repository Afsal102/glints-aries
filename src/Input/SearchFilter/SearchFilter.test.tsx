import * as React from 'react';

import * as renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import SearchFilter from '.';

interface Props {
  label: string;
  onChange: jest.Mock;
  content: React.ReactNode;
}

const props: Props = {
  label: 'Search jobs, location, or company',
  onChange: jest.fn().mockImplementation(event => event.target.value),
  content: (
    <SearchFilter.Button variant="solid-blue">Search</SearchFilter.Button>
  ),
};
const lists = [
  { title: 'Popular Careers', item: 'Software Engineering' },
  { title: 'Popular Locations', item: 'Singapore' },
  { title: 'Popular Companies', item: 'Glints' },
];
const inputValue = 'glints';

function setupSearchFilter(props: Props) {
  const { getByPlaceholderText, getByRole, ...utils } = render(
    <SearchFilter
      label={props.label}
      onChange={props.onChange}
      content={props.content}
    >
      <SearchFilter.Body>
        {lists.map(list => (
          <SearchFilter.List title={list.title} key={list.title}>
            <SearchFilter.Item>{list.item}</SearchFilter.Item>
          </SearchFilter.List>
        ))}
      </SearchFilter.Body>
    </SearchFilter>
  );
  const searchFilterNode = getByRole('search') as HTMLDivElement;
  const searchFilterInput = getByPlaceholderText(
    props.label
  ) as HTMLInputElement;
  const searchFilterMenu = getByRole('menuitem', {
    hidden: true,
  }) as HTMLDivElement;
  return { searchFilterNode, searchFilterInput, searchFilterMenu, utils };
}

function setupSearchFilterWithChangeEvent(props: Props) {
  const { searchFilterInput } = setupSearchFilter(props);
  userEvent.type(searchFilterInput, inputValue);
  return { searchFilterInput };
}

function createNodeMock(element: any) {
  if (element.type === 'div') {
    return {
      contains: jest.fn(),
    };
  }
  return null;
}

it('<SearchFilter> should render with an input with a button, and a list with an item', () => {
  const options = { createNodeMock };
  const SearchFilterSnapshot = renderer
    .create(
      <SearchFilter
        label={props.label}
        onChange={props.onChange}
        content={props.content}
      >
        <SearchFilter.List title="Title">
          <SearchFilter.Item>Item</SearchFilter.Item>
        </SearchFilter.List>
      </SearchFilter>,
      options
    )
    .toJSON();
  expect(SearchFilterSnapshot).toMatchSnapshot();
});

describe('when it is rendered', () => {
  it('should display the correct placeholder', () => {
    const { searchFilterInput } = setupSearchFilter(props);
    expect(searchFilterInput).toBeTruthy();
  });

  it('should have an empty string as its value', () => {
    const { searchFilterInput } = setupSearchFilter(props);
    expect(searchFilterInput.value).toEqual('');
  });

  it('should not display the menu', () => {
    const { searchFilterMenu } = setupSearchFilter(props);
    expect(searchFilterMenu).not.toHaveAttribute('open');
  });

  it('should display the correct content prop', () => {
    const { searchFilterNode, utils } = setupSearchFilter(props);
    const searchFilterContent = utils.getByText('Search');
    expect(searchFilterNode).toContainElement(searchFilterContent);
  });

  lists.forEach(list => {
    it(`should display the correct list "${list.title}" and item "${list.item}"`, () => {
      const { utils } = setupSearchFilter(props);
      expect(utils.getByText(list.title)).toBeTruthy();
      expect(utils.getByText(list.item)).toBeTruthy();
    });
  });
});

describe('when it is focused', () => {
  it('should open menu', () => {
    const { searchFilterInput, searchFilterMenu } = setupSearchFilter(props);
    fireEvent.focus(searchFilterInput);
    expect(searchFilterMenu).toHaveAttribute('open');
  });

  it('should close when clicking outside the Search Filter', () => {
    const { searchFilterInput, searchFilterMenu, utils } = setupSearchFilter(
      props
    );
    fireEvent.focus(searchFilterInput);
    fireEvent.click(utils.container);
    expect(searchFilterMenu).not.toHaveAttribute('open');
  });

  lists.forEach(list => {
    it('should close when clicking on each list item', () => {
      const { searchFilterInput, searchFilterMenu, utils } = setupSearchFilter(
        props
      );
      const listItem = utils.getByText(list.item);
      fireEvent.focus(searchFilterInput);
      fireEvent.click(listItem);
      expect(searchFilterMenu).not.toHaveAttribute('open');
    });
  });
});

describe('when a value is entered', () => {
  it('should call onChange for each input', () => {
    setupSearchFilterWithChangeEvent(props);
    expect(props.onChange).toHaveBeenCalledTimes(inputValue.length);
  });

  it('should return the input value when onChange is passed a function: event => event.target.value', () => {
    setupSearchFilterWithChangeEvent(props);
    expect(props.onChange.mock.results.slice(-1)[0].value).toEqual(inputValue);
  });

  it('should display the correct value', () => {
    const { searchFilterInput } = setupSearchFilterWithChangeEvent(props);
    expect(searchFilterInput.value).toEqual(inputValue);
  });
});

describe('<SearchFilter> inputRef', () => {
  test('inputRef is being correctly passed to the underlying input element', () => {
    const label = 'test-inputRef';
    const inputRef = React.createRef<HTMLInputElement>();
    const { queryByPlaceholderText } = render(
      <SearchFilter label={label} inputRef={inputRef} />
    );
    expect(queryByPlaceholderText(label)).toEqual(inputRef.current);
  });

  test('inputRef can focus the underlying input element programmatically', () => {
    const label = 'test-inputRef';
    const inputRef = React.createRef<HTMLInputElement>();
    const { queryByPlaceholderText } = render(
      <SearchFilter label={label} inputRef={inputRef} />
    );
    inputRef.current.focus();
    expect(queryByPlaceholderText(label)).toHaveFocus();
  });
});
