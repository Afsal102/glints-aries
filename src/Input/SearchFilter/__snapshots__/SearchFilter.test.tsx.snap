// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchFilter> should render with an input with a button, and a list with an item 1`] = `
<div
  aria-expanded={false}
  aria-label="Search jobs, location, or company"
  className="SearchFilterStyle__SearchFilterContainer-sc-18veyec-0 bUBbHi aries-searchfilter"
  role="search"
>
  <div
    className="SearchFilterStyle__SearchFilterBar-sc-18veyec-1 eHOovt searchfilter-inputwrapper"
  >
    <input
      onChange={[MockFunction]}
      onFocus={[Function]}
      placeholder="Search jobs, location, or company"
      type="text"
    />
    <div
      className="ButtonStyle__SolidBtnContainer-gw4ytw-2 ggZAqT aries-solid-btn SearchFilterStyle__SearchFilterButton-sc-18veyec-2 dvMPkA"
    >
      <button
        className="ButtonStyle__Button-gw4ytw-0 ButtonStyle__SolidBtn-gw4ytw-1 iJVRTu solid-btn-content"
      >
        Search
      </button>
    </div>
  </div>
  <div
    aria-hidden={true}
    className="SearchFilterStyle__SearchFilterBodyWrapper-sc-18veyec-3 aDmLC searchfilter-content"
    open={false}
    role="menuitem"
  >
    <div
      className="SearchFilterStyle__SearchFilterListWrapper-sc-18veyec-5 jMzbtT searchfilter-listitem"
      role="list"
    >
      <label>
        Title
      </label>
      <ul>
        <li
          className="SearchFilterStyle__SearchFilterItemWrapper-sc-18veyec-6 fzkjMH"
          onClick={[Function]}
          role="listitem"
        >
          Item
        </li>
      </ul>
    </div>
  </div>
</div>
`;
